runtime/src/update_did_signed_extension.rs

```diff
@@ -0,0 +1,160 @@
+use crate::{identity, identity::LinkedKeyInfo, runtime, Runtime};
+use primitives::{IdentityId, Key, TransactionError};
+
+use sr_primitives::{
+    traits::SignedExtension,
+    transaction_validity::{
+        InvalidTransaction, TransactionValidity, TransactionValidityError, ValidTransaction,
+    },
+};
+
+use codec::{Decode, Encode};
+use core::convert::TryFrom;
+use rstd::marker::PhantomData;
+use srml_support::dispatch::DispatchInfo;
+
+type Identity = identity::Module<Runtime>;
+type Call = runtime::Call;
+
+/// This signed extension double-checks and updates the current identifier extracted from
+/// the caller account in each transaction.
+#[derive(Default, Encode, Decode, Clone, Eq, PartialEq)]
+pub struct UpdateDid<T: system::Trait + Send + Sync>(PhantomData<T>);
+
+impl<T: system::Trait + Send + Sync> UpdateDid<T> {
+    pub fn new() -> Self {
+        UpdateDid(PhantomData)
+    }
+
+    /// It extracts the `IdentityId` associated with `who` account.
+    fn identity_from_key(who: &T::AccountId) -> Option<IdentityId> {
+        if let Ok(who_key) = Key::try_from(who.encode()) {
+            if let Some(linked_key_info) = Identity::key_to_identity_ids(&who_key) {
+                if let LinkedKeyInfo::Unique(id) = linked_key_info {
+                    return Some(id);
+                }
+            }
+        }
+        None
+    }
+}
+
+#[cfg(feature = "std")]
+impl<T: system::Trait + Send + Sync> rstd::fmt::Debug for UpdateDid<T> {
+    fn fmt(&self, f: &mut rstd::fmt::Formatter) -> rstd::fmt::Result {
+        write!(f, "UpdateDid")
+    }
+}
+
+impl<T: system::Trait + Send + Sync> SignedExtension for UpdateDid<T> {
+    type AccountId = T::AccountId;
+    type Call = runtime::Call;
+    type AdditionalSigned = ();
+    type Pre = ();
+
+    fn additional_signed(&self) -> rstd::result::Result<(), TransactionValidityError> {
+        Ok(())
+    }
+
+    /// It ensures that transaction caller account has an associated identity and
+    /// that identity has been validated by any KYC.
+    /// The current identity will be accesible through `Identity::current_did`.
+    ///
+    /// Only the following methods can be called with no identity:
+    ///     - `identity::register_did`
+    fn validate(
+        &self,
+        who: &Self::AccountId,
+        call: &Self::Call,
+        _: DispatchInfo,
+        _: usize,
+    ) -> TransactionValidity {
+        match call {
+            // Add here any function from any module which does *not* need a current identity.
+            Call::Identity(identity::Call::register_did(..)) => Ok(ValidTransaction::default()),
```

Comment:
Wouldn't it be more convenient to pattern match calls that require KYC check and let the rest fall through? That way we'd only need to edit this file when DID is needed for an AccountID as opposed to editing the file whenever a new module/call is added.

Reply:
Resolved at slack chat.